lib/tesla/middleware/telemetry.ex

```diff
@@ -3,30 +3,98 @@ if Code.ensure_loaded?(:telemetry) do
     @behaviour Tesla.Middleware
 
     @moduledoc """
-    Send the request time and meta-information through telemetry.
+    Emits events using the `:telemetry` library to expose instrumentation.
 
     ### Example usage
     ```
     defmodule MyClient do
       use Tesla
 
-      plug Tesla.Middleware.Telemetry
+      plug Tesla.Middleware.Telemetry, prefix: [:my_client]
 
     end
 
-    :telemetry.attach("my-tesla-telemetry", [:tesla, :request], fn event, time, meta, config ->
+    :telemetry.attach("my-tesla-telemetry", [:my_client, :tesla, :request, stop], fn event, measurements, meta, config ->
       # Do something with the event
     end)
     ```
 
+    ### Options
+
+    * `:event_prefix` - a list of atoms to prefix to the telemetry event. Defaults to `[]`
+
+    ## Telemetry Events
+
+    * `[:tesla, :request, :start]` - emitted at the beginning of the request.
+      * Measurement: `%{time: System.monotonic_time}`
+      * Metadata: `%{env: Tesla.Env.t}`
+
+    * `[:tesla, :request, :stop]` - emitted at the end of the request.
+      * Measurement: `%{duration: native_time}`
+      * Metadata: `%{env: Tesla.Env.t}`
+
+    * `[:tesla, :request, :error]` - emitted at the end of the request when there is an error.
+      * Measurement: `%{duration: native_time}`
+      * Metadata: `%{env: Tesla.Env.t, reason: term}`
+
+    * `[:tesla, :request, :exception]` - emitted at the end of the request when an exception is raised.
+      * Measurement: `%{duration: native_time}`
+      * Metadata: `%{env: Tesla.Env.t, exception: Exception.t, stacktrace: Exception.stacktrace}`
+
     Please check the [telemetry](https://hexdocs.pm/telemetry/) for the further usage.
     """
 
     @doc false
-    def call(env, next, _opts) do
-      {time, res} = :timer.tc(Tesla, :run, [env, next])
-      :telemetry.execute([:tesla, :request], %{request_time: time}, %{result: res})
-      res
+    def call(req_env, next, opts) do
+      prefix = Keyword.get(opts, :event_prefix, [])
+      start_time = System.monotonic_time()
+
+      try do
+        emit_start(req_env, start_time, prefix)
+
+        Tesla.run(req_env, next)
+        |> emit_result(start_time, prefix, req_env)
+      rescue
+        e ->
+          stacktrace = System.stacktrace()
+          metadata = %{env: req_env, exception: e, stacktrace: stacktrace}
+
+          :telemetry.execute(
+            prefix ++ [:tesla, :request, :exception],
+            %{duration: System.monotonic_time() - start_time},
+            metadata
+          )
+
+          reraise e, stacktrace
+      end
+    end
+
+    defp emit_start(req_env, start_time, prefix) do
+      :telemetry.execute(prefix ++ [:tesla, :request, :start], %{time: start_time}, %{
+        env: req_env
+      })
+    end
+
+    defp emit_result(result, start_time, prefix, req_env) do
+      try do
+        result
```

Comment:
What is this supposed to do? 🤔

Reply:
I was avoiding nesting two of these in the main function. :)

Since `after` is evaluated after the try or catch bodies are evaluated, recording the end event after those bodies have run gives us the actual time that the operation took. Since it is possible for Tesla to raise an exception, the first `try do` in the main function would emit the event with the `env` being what was passed to the function (we can't read evaluated values from the `try do` block because it may not exist in the case of an exception being raised. 

In the `emit_result` function I was using the same pattern to emit the event as a side effect. This may not be absolutely necessary since we're just returning the value passed, however it wouldn't cause any harm, so I _think_ it would be ok to omit it if you have strong feelings about it.

Reply:
Elixir is a strict language - arguments are evaluated before they are passed to functions. This `try` will not have any effect.

Reply:
Correct. Normally we do everything in the `after` as a little trick but there's no guarantee the result would have been evaluated since `after` runs on success or on exception. Either way, this is a moot point with the changes I made last night to simplify it. 👍